src/components/FileUpload/FileUpload.tsx

```diff
@@ -0,0 +1,205 @@
+import React, { useCallback, useEffect, useState } from 'react'
+import { useDropzone } from 'react-dropzone'
+
+import uploadIcon from '../../images/file-upload.svg'
+import { Button } from '../Button'
+import FormControlWarningError from '../common/FormControlWarningError'
+import { Icon } from '../Icon'
+import FilePreview, { FileWithPreview } from './components/FilePreview'
+import PasswordModal from './components/PasswordModal'
+import {
+	Label,
+	LabelDesktop,
+	LabelTouchDevice,
+	PreviewFilesWrapper,
+	RelativeWrap,
+	Remove,
+	StyledUpload,
+	UploadIcon,
+	Wrapper
+} from './FileUpload.styles'
+import { FileUploadProps, PasswordModalFile } from './FileUpload.types'
+
+export const FileUpload: React.FC<FileUploadProps> = ({
+	multiple = true,
+	additive = false,
+	testId = 'FileUpload',
+	files,
+	setFiles: _setFiles,
+	label,
+	labelTouchDevice,
+	dragInstructions,
+	onFileDrop,
+	acceptedFilePattern,
+	uploadButtonText,
+	uploadButtonTextTouchDevice,
+	onFileRemove,
+	error,
+	className,
+	fileValidation,
+	withIcon,
+	onPassword,
+	passwordPromptText = 'Enter the password to open this PDF file.',
+	passwordIncorrectText = 'Invalid password. Please try again.',
+	passwordConfirmButtonText = 'Confirm',
+	passwordCancelButtonText = 'Cancel',
+	passwordPlaceholderText = 'Password'
+}) => {
+	const [internalErrors, setInternalErrors] = useState<string[]>([])
+	const [passwordModalQueue, setPasswordModalQueue] = useState<
+		PasswordModalFile[]
+	>([])
```

Comment:
What about creating the custom hook for the queue operations?

It could work something like this:
`usePasswordQueue` hook that returns the following object:
{
  `currentFile`: `PasswordModalFile` or null if queue is empty
  `setAttr`: `(fileIdentificator, attrs) => void`
  `addItemToQueue`: `(file) => void`
  `editItemInQueue`: edits item in the queue and moves this item to the start of the queue
}

Take this just as the recommendation. I'm used to writing the custom hooks like this, because all the logic is in one place and it's easier to navigate the code imo.

Reply:
Thanks very much for this suggestion. The code looks much cleaner now 😊 I made some changes against your original idea but hopefully it's more or less as you intended. Please take a look.